### PR TITLE
[BEAM-1231] Use well known coder types in Java

### PR DESCRIPTION
**Pull request by lukecwik**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @dhalperi 
R: @vikkyrk

**Comment by dhalperi** [comment]
retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java [hunk 1]
```diff
@@ -147,6 +151,10 @@ public CloudObject asCloudObject() {
     return result;
   }
 
+  protected CloudObject initializeCloudObject() {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java]: javadoc?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/LengthPrefixCoder.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+  @Override
+  protected CloudObject initializeCloudObject() {
+    return CloudObject.forClassName("kind:length_prefix");
+  }
+
+  @Override
+  public void encode(T value, OutputStream outStream, Context context)
+      throws CoderException, IOException {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/LengthPrefixCoder.java]: A bunch of questions about LPC:

this ignores the case when `context` is outer context. Probably deliberate, but should be documented?

If this is a cross-SDK whose encoding is standardized, should be put spec-like comments in the doc explaining exactly what certain values should encode to?

Should we link to the specific `VarInt` implementation?

Should we use `VarLong` for values longer than 2GB?

Is there any way to document which coders are "fixed"? I could imagine changing VarIntCoder to some more efficient VarInt format if one exists, like the one we discussed yesterday.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/GlobalWindowTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
+/*
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/windowing/GlobalWindowTest.java]: nice

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/LengthPrefixCoderTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+  /**
+   * Generated data to check that the wire format has not changed. To regenerate, see
+   * {@link org.apache.beam.sdk.coders.PrintBase64Encodings}.
+   */
+  private static final List<String> TEST_ENCODINGS = ImmutableList.of(
+      "AwoLDA",
+      "Ag0D",
+      "Ag0O",
+      "AA");
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/LengthPrefixCoderTest.java]: put it near `TEST_VALUES`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/LengthPrefixCoder.java (hunk 2 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/LengthPrefixCoder.java]: This change is about using kind:* and doesn't get into actually making the binary encodings equivalent across Python/Java/...

A future change will address it for well known coder types using test files that store the encoded and decoded versions plus the documentation required.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java (hunk 1 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/LengthPrefixCoderTest.java (hunk 4 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/LengthPrefixCoderTest.java]: Done

**Comment by lukecwik** [comment]
Example failing job: 2016-12-29_16_48_23-4024298823312880713

java.lang.RuntimeException: Unable to deserialize class interface org.apache.beam.sdk.coders.Coder
	at org.apache.beam.sdk.util.Serializer.deserialize(Serializer.java:102)
	at com.google.cloud.dataflow.worker.runners.worker.MapTaskExecutorFactory$4.typedApply(MapTaskExecutorFactory.java:436)
	at com.google.cloud.dataflow.worker.runners.worker.MapTaskExecutorFactory$4.typedApply(MapTaskExecutorFactory.java:431)
	at com.google.cloud.dataflow.worker.graph.Networks$TypeSafeNodeFunction.apply(Networks.java:55)
	at com.google.cloud.dataflow.worker.graph.Networks$TypeSafeNodeFunction.apply(Networks.java:43)
	at com.google.cloud.dataflow.worker.graph.Networks.replaceDirectedNetworkNodes(Networks.java:78)
	at com.google.cloud.dataflow.worker.runners.worker.MapTaskExecutorFactory.create(MapTaskExecutorFactory.java:123)
	at ...
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Invalid type id 'kind:pair' (for id type 'Id.class'): no such class found (through reference chain: org.apache.beam.sdk.util.FullWindowedValueCoder["component_encodings"]->java.util.ArrayList[0]->com.google.cloud.dataflow.sdk.util.TimerOrElementCoder["component_encodings"]->java.util.ArrayList[0])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:339)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:311)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:298)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:249)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:26)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:490)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:465)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:380)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1123)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:298)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:168)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:135)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:120)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:91)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:142)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:279)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:249)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:26)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:490)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:465)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:380)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1123)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:298)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:168)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:135)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:120)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:91)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:142)
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:42)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3761)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2042)
	at com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:2529)
	at org.apache.beam.sdk.util.Serializer.deserialize(Serializer.java:97)
	... 12 more
Caused by: java.lang.IllegalArgumentException: Invalid type id 'kind:pair' (for id type 'Id.class'): no such class found
	at com.fasterxml.jackson.databind.jsontype.impl.ClassNameIdResolver._typeFromId(ClassNameIdResolver.java:66)
	at com.fasterxml.jackson.databind.jsontype.impl.ClassNameIdResolver.typeFromId(ClassNameIdResolver.java:48)
	at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._findDeserializer(TypeDeserializerBase.java:158)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:106)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:91)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromAny(AsPropertyTypeDeserializer.java:163)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserializeWithType(UntypedObjectDeserializer.java:554)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:279)
	... 42 more

**Comment by lukecwik** [comment]
@dhalperi I added a bunch of tests to make sure well known coder types when used as component coders are able to be deserialized. Would you like to review the additional diffs?

**Comment by dhalperi** [comment]
LGTM

**Pull request closed without merging**